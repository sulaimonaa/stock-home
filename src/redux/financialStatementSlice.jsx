import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const API_KEY = '2a983b252a14dce3b861a39db579c57f';

export const FetchFinancialStatements = createAsyncThunk(
  'financialstatements/fetchstatement',
  async () => {
    try {
      const response = await fetch(
        `https://financialmodelingprep.com/api/v3/financial-statement-symbol-lists?apikey=${API_KEY}`
      );
      if (!response.ok) {
        throw new Error('Network response was not ok');
      }
      const data = await response.json();
      return data;
    } catch (error) {
      throw new Error('Something went wrong with fetching data');
    }
  }
);

const initialState = {
  value: [],
  status: 'idle',
  error: false,
};

const financialStatementReducer = createSlice({
  name: 'financialstatements',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(FetchFinancialStatements.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(FetchFinancialStatements.fulfilled, (state, action) => ({
        ...state,
        status: 'successful',
        value: action.payload,
      }))
      .addCase(FetchFinancialStatements.rejected, (state, action) => ({
        ...state,
        status: 'failed',
        error: action.error.message,
      }))
  },
});

export default financialStatementReducer.reducer;
