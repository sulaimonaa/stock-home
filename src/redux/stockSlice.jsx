import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const API_KEY = '2a983b252a14dce3b861a39db579c57f';

export const FetchStockLists = createAsyncThunk(
  'stocks/fetchstocks',
  async () => {
    try {
      const response = await fetch(
        `https://financialmodelingprep.com/api/v3/stock/list?apikey=${API_KEY}`
      );
      if (!response.ok) {
        throw new Error('Network response was not ok');
      }
      const data = await response.json();
      return data;
    } catch (error) {
      throw new Error('Something went wrong with fetching data');
    }
  }
);

const initialState = {
  value: [],
  status: 'idle',
  error: false,
};

const stockListReducer = createSlice({
  name: 'stocks',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(FetchStockLists.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(FetchStockLists.fulfilled, (state, action) => ({
        ...state,
        status: 'successful',
        value: action.payload,
      }))
      .addCase(FetchStockLists.rejected, (state, action) => ({
        ...state,
        status: 'failed',
        error: action.error.message,
      }))
  },
});

export default stockListReducer.reducer;
