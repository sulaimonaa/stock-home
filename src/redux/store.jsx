import { configureStore, getDefaultMiddleware } from "@reduxjs/toolkit";
import thunk from "redux-thunk";
import financialStatementReducer from './financialStatementSlice';
import stockListReducer from './stockSlice'

const store = configureStore({
    reducer: {
        financialstatements: financialStatementReducer,
        stocks: stockListReducer
    },
    middleware: [thunk, ...getDefaultMiddleware()]
});

export default store;
