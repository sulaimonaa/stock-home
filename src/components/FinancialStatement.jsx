import React, { useEffect, useState } from 'react'; // eslint-disable-line no-undef, no-unused-vars
import { useDispatch, useSelector } from 'react-redux';
import { FetchFinancialStatements } from '../redux/financialStatementSlice';
import FinancialStatementList from './FinancialStatementList';
import { mapFinancialToValue } from './FinancialMapping';
import Navigation from './OtherNav';
import '../styles/financiallist.css';

const FinancialStatement = () => {
  const dispatch = useDispatch();
  const financialList = useSelector((state) => state.financialstatements.value);

  const filterAlphabetic = (dataList) => {
    return dataList.filter(item => /^[a-zA-Z]/.test(item));
  };

  const [searchTerm, setSearchTerm] = useState('');
  const [searchedData, setSearchedData] = useState([]);

  useEffect(() => {
    dispatch(FetchFinancialStatements());
  }, [dispatch]);

  useEffect(() => {
    const filteredData = filterAlphabetic(financialList);
    const limitedData = filteredData.slice(0, 5000);

    const filtered = limitedData.filter((stock) =>
      mapFinancialToValue(stock).toLowerCase().includes(searchTerm.toLowerCase())
    );
    setSearchedData(filtered);
  }, [financialList, searchTerm]);

  return (
    <>
    <Navigation />
    <div className='wrap'>
      <h1 className='list-h1'>Financial statements list</h1>
      <input className='search' value={searchTerm}
        onChange={(e) => setSearchTerm(e.target.value)} placeholder='Search Company by symbol' />
      <ul>
        {searchedData.map((list, index) => (
          <FinancialStatementList key={index} id={index + 1} symbolName={mapFinancialToValue(list)} />
        ))}
      </ul>
    </div>
    </>
  );
};

export default FinancialStatement;