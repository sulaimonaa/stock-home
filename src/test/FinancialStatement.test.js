import React from 'react'; // eslint-disable-line no-undef, no-unused-vars
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux'; 
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import FinancialStatement from '../components/FinancialStatement';
import { FetchFinancialStatements } from '../redux/financialStatementSlice';
import { BrowserRouter } from 'react-router-dom';

const mockFinancialList = [
  "AAP", "AAPL", "CAC", "EEC", "ACA", "ACP"
];
/* eslint-disable no-undef, no-unused-vars */
describe('FinancialStatement', () => {
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  
  it('fetches financial statements and displays them', async () => {
    const store = mockStore({
      financialstatements: { value: mockFinancialList },
    });

    // Dispatch the async action using store.dispatch
    await store.dispatch(FetchFinancialStatements());

    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <FinancialStatement />
        </BrowserRouter>
      </Provider>
    );

    // Wait for API call and component updates
    await screen.findAllByTestId('financial-statement-list');

    // Assert and update snapshot
    expect(asFragment()).toMatchSnapshot();
  });

  it('filters financial statements based on search term', async () => {
    const store = mockStore({
      financialstatements: { value: mockFinancialList },
    });

    // Dispatch the async action using store.dispatch
    await store.dispatch(FetchFinancialStatements());

    render(
      <Provider store={store}>
        <BrowserRouter>
          <FinancialStatement />
        </BrowserRouter>
      </Provider>
    );

    // Wait for API call and component updates
    await screen.findAllByTestId('financial-statement-list');

    // Simulate user input
    fireEvent.change(screen.getByPlaceholderText('Search Company by symbol'), {
      target: { value: 'AAPL' },
    });

    await screen.findByTestId('financial-statement-list');

    const displayedFinancialStatements = screen.getAllByTestId('financial-statement-list');
    expect(displayedFinancialStatements).toHaveLength(1);

    expect(displayedFinancialStatements[0]).toMatchSnapshot();
  });

});
