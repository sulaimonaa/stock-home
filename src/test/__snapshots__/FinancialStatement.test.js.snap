// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinancialStatement fetches financial statements and displays them 1`] = `
<DocumentFragment>
  <div
    class="nav"
  >
    <div
      class="container navbar"
    >
      <div
        class="home-icon"
      >
        <img
          alt="previous icon"
          src="mocked-file-url"
        />
      </div>
      <div
        class="home-icon"
      >
        <a
          aria-current="page"
          class="active"
          href="/"
        >
          <img
            alt="home icon"
            src="mocked-file-url"
          />
        </a>
      </div>
      <div
        class="home-icon"
      >
        <a
          class=""
          href="/stocks"
        >
          <img
            alt="home icon"
            src="mocked-file-url"
          />
        </a>
      </div>
      <div
        class="menu-icon"
      >
        <a
          class=""
          href="/about"
        >
          <img
            alt="menu icon"
            src="mocked-file-url"
          />
        </a>
      </div>
    </div>
  </div>
  <div
    class="wrap"
  >
    <h1
      class="list-h1"
    >
      Financial statements list
    </h1>
    <input
      class="search"
      placeholder="Search Company by symbol"
      value=""
    />
    <ul>
      <li
        data-testid="financial-statement-list"
        id="1"
      >
        <a
          class="financial-list"
          href="/company/AAP"
        >
          AAP
        </a>
      </li>
      <li
        data-testid="financial-statement-list"
        id="2"
      >
        <a
          class="financial-list"
          href="/company/AAPL"
        >
          AAPL
        </a>
      </li>
      <li
        data-testid="financial-statement-list"
        id="3"
      >
        <a
          class="financial-list"
          href="/company/CAC"
        >
          CAC
        </a>
      </li>
      <li
        data-testid="financial-statement-list"
        id="4"
      >
        <a
          class="financial-list"
          href="/company/EEC"
        >
          EEC
        </a>
      </li>
      <li
        data-testid="financial-statement-list"
        id="5"
      >
        <a
          class="financial-list"
          href="/company/ACA"
        >
          ACA
        </a>
      </li>
      <li
        data-testid="financial-statement-list"
        id="6"
      >
        <a
          class="financial-list"
          href="/company/ACP"
        >
          ACP
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`FinancialStatement filters financial statements based on search term 1`] = `
<li
  data-testid="financial-statement-list"
  id="1"
>
  <a
    class="financial-list"
    href="/company/AAPL"
  >
    AAPL
  </a>
</li>
`;
